### PR TITLE
Add deprecation notice to adhoc test precondition

### PR DESCRIPTION
**Pull request by ov7a**
Also, add a note regarding multiple toolchains.

follow up to #23428

### CODE REVIEW

File: subprojects/internal-testing/src/main/groovy/org/gradle/util/Requires.java [hunk 1]
```diff
@@ -32,5 +32,12 @@
 public @interface Requires {
     TestPrecondition[] value() default {TestPrecondition.NULL_REQUIREMENT};
 
+    /**
+     * @deprecated Do not introduce new adhoc logic.
+     * Consider avoiding it entirely or adding a new TestPrecondition enum value instead.
+     * It is hard to ensure a test with this annotation would be run at CI at least once.
+     * See <a href="https://github.com/gradle/gradle-private/issues/3616">issue</a> for more details.
+     */
+    @Deprecated
```
Comment (wolfs) [inline subprojects/internal-testing/src/main/groovy/org/gradle/util/Requires.java:41]: 🤔 While this makes sense to deprecate adding more ad-hoc logic, when we introduce this here we'd also need to annotate all the places that use it now to ignore the deprecation warning, right? I wonder if that is worth it.

Comment (ljacomet) [reply subprojects/internal-testing/src/main/groovy/org/gradle/util/Requires.java:41]: I think adding this is a step in the right direction and we indeed already have deprecated test helpers in other places. So +1 from me.

**Review by ljacomet — APPROVED** [review]

**Comment by ov7a** [comment]
@bot-gradle test and merge

**Comment by ov7a** [comment]
@bot-gradle test and merge

**Pull request merged**